File: solutions/leetcode/377.组合总和-ⅳ.py
#
# @lc app=leetcode.cn id=377 lang=python3
#
# [377] 组合总和 Ⅳ
#

from typing import *
from collections import *

# @lc code=start
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        nums.sort()
        dp = [0 for _ in range(target + 1)]
        for num in nums:
            if num <= target:
                dp[num] += 1
        
        for i in range(target + 1):
            for num in nums:
                if num > i:
                    break
                dp[i] += dp[i - num]
        return dp[target]
        
# @lc code=end

s = Solution()
nums = list(map(int, input().split()))
target = int(input())
print(s.combinationSum4(nums, target))

File: solutions/leetcode/377.组合总和-ⅳ.cpp
/*
 * @lc app=leetcode.cn id=377 lang=cpp
 *
 * [377] 组合总和 Ⅳ
 */

// @lc code=start
/*
用f[i]表示构成数字i的方案数
如果i属于nums，初始值为1，否则为0
f[i] += f[i - num]
*/
class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<long long> f(target + 1, 0);
        sort(nums.begin(), nums.end());
        const long long MOD = INT_MAX + 1;
        for (int i = 1; i <= target; i++) {
            for (auto& num : nums) {
                if (num == i) {
                    f[i]++;
                    f[i] %= MOD;
                }
                if (i > num) {
                    f[i] += f[i - num];
                    f[i] %= MOD;
                }
                else
                    break;
            }
        }
        // for (auto & num : f)
        //     cout << num << ' ';
        return f[target];
    }
};
// @lc code=end

